In [12]:
!pip install mysqlclient

In [14]:
!pip install nbmerge

  Preparing metadata (setup.py) ... done
  Created wheel for nbmerge: filename=nbmerge-0.0.4-py2.py3-none-any.whl size=6393 sha256=3f1b3f261a9c91988dce9cbf5941c96e836fa714ae74460cf3046968e1f371cc
  Stored in directory: /Users/MichaelAlberts/Library/Caches/pip/wheels/0b/a8/31/4ae563f674b488c82ee72cfe0f86487c21532db97ae910b264
Successfully built nbmerge


In [16]:
!nbmerge data_collection.ipynb sql_analysis.ipynb -o presentation.ipynb

In [13]:
!pip install sqlalchemy==1.4.4

In [11]:
%load_ext sql

In [5]:
%sql mysql://USER:PASSWORD@DATABASE

# Exploratory Question #1
We can use this query to find out how often our capsules are able to be reused.

In [6]:
%%sql
  SELECT 
    COUNT(*) as num_capsules, 
    SUM(reuse_count) as total_reuse_count
  FROM capsules;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
1 rows affected.


[(19, Decimal('9'))]

This query shows that, of our 19 capsules used, 9 of them were able to be reused again in other missions.

# Exploratory Question #2
This query is able to keep track of our launches as well as what their current statuses are.

In [7]:
%%sql
  SELECT 
    COUNT(*) as num_launches, status, launch_success
  FROM launches
  LEFT JOIN capsules 
    ON launches.flight_number = capsules.flight_number
  GROUP BY status, launch_success
  ORDER BY num_launches DESC;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
7 rows affected.


[(88, None, 1),
 (9, 'active', 1),
 (4, None, 0),
 (4, 'unknown', 1),
 (3, None, None),
 (2, 'retired', 1),
 (1, 'destroyed', 0)]

The above query shows the results of all of our launches.  Showing the amount of launches where the ship or capsule launched or failed.  As well as what its current status is.

# Exploratory Question #3
This query is used to find the payloads with the highest mass.

In [8]:
%%sql
  SELECT 
    payload_id, payload_mass_lbs
  FROM payloads
  ORDER BY payload_mass_lbs DESC
  LIMIT 10;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
10 rows affected.


[('Starlink 5', 34392.113),
 ('Starlink 3', 34392.113),
 ('Starlink 1', 34392.1),
 ('Starlink 6', 33951.2),
 ('Starlink-10', 33951.2),
 ('Starlink 4', 33951.2),
 ('Starlink-14', 33951.2),
 ('Starlink-13', 33951.2),
 ('Starlink-12', 33951.2),
 ('Starlink 2', 33951.2)]

This query is used to find out which payload per launch was the heaviest.  This shows that our starlink launches are currently the heaviest.

# Exploratory Question #4
This query is used to easliy keep track of which ships are currently active versus inactive.

In [9]:
%%sql
  SELECT COUNT(*) as num_ships, active
  FROM ships
  GROUP BY active;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
2 rows affected.


[(12, 0), (10, 1)]

This query keeps track of the number of starships we currently have active (1) versus the currently inactive ships (0).

# Exploratory Question #5
This query is used to show the number of missions that each capsule has gone on.

In [10]:
%%sql
  SELECT 
    capsule_id, COUNT(*) as num_missions
  FROM capsules
  GROUP BY capsule_id;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
2 rows affected.


[('dragon1', 13), ('dragon2', 6)]

This query is used to keep track of the amount of missions that our two capsule types have undergone.

# Business Question #1
How many launches have been successful, and what is the success rate over time?
#### Business Value:
This business question is valuable for assessing the overall success rate of SpaceX launches and identifying factors that may affect the success rate.  The insights gained from this question can be used to improve the reliability of future launches and potentially reduce costs associated with failed launches.

#### SQL Functions
This query utilizes a View, Case and Group By.

In [42]:
%%sql
  CREATE VIEW launch_stats AS
  SELECT
    launch_year,
    COUNT(*) AS num_launches,
    SUM(CASE WHEN launch_success = 1 THEN 1 ELSE 0 END) AS num_successes
  FROM launches
  GROUP BY launch_year;
  SELECT
    launch_year,
    num_launches,
    ROUND(num_successes / num_launches * 100) AS success_rate
  FROM launch_stats
  ORDER BY launch_year;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
0 rows affected.
14 rows affected.


[('2006', 1, Decimal('0')),
 ('2007', 1, Decimal('0')),
 ('2008', 2, Decimal('50')),
 ('2009', 1, Decimal('100')),
 ('2010', 2, Decimal('100')),
 ('2012', 2, Decimal('100')),
 ('2013', 3, Decimal('100')),
 ('2014', 6, Decimal('100')),
 ('2015', 7, Decimal('86')),
 ('2016', 9, Decimal('89')),
 ('2017', 18, Decimal('100')),
 ('2018', 21, Decimal('100')),
 ('2019', 13, Decimal('100')),
 ('2020', 25, Decimal('88'))]

### Insight and Reccomendation
This query shows the mission success rate over the course of fourteen years at SpaceX.  SpaceX has demonstrated an extremely high success rate in its launches.  My recommendation would be for the company to continue its focus on quality and safety in its launch operations, while also exploring new opportunities to expand its capabilities and offerings in the space industry

## Sub-Question 1
How many missions were flown using capsules with a reuse count greater than 1?

#### Business Value
Some business justifications for this query are; to evaluate the effectiveness and efficiency of SpaceX's reuse strategy for capsules, help assess the cost savings and resource efficiency of SpaceX's capsule reuse strategy.

#### SQL Functions
This query utilizes a the join and subquery function.

In [30]:
%%sql
  SELECT 
    COUNT(DISTINCT mission_name) AS mission_count
  FROM launches
  WHERE flight_number IN (
    SELECT 
      l.flight_number
    FROM capsules c
    JOIN launches l ON c.flight_number = l.flight_number
    WHERE c.reuse_count >= 1
  );

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
1 rows affected.


[(6,)]

### Insight and Reccomendation
This query shows that only six missions were undertaken with a re-used capsule.  A reccomendation would be to analyze what worked with those six missions and to attempt to replicate the process for future missions in order to potentially reduce overall costs for re-making capsules.

## Sub-Question 2
What is the average and median reuse count for each capsule type?

#### Business Value
This information can help the project managers to identify the most durable and reliable capsule types, which can result in cost savings by reducing the number of new capsules required to be manufactured or purchased.  This data can also help in identifying any issues or inefficiencies with the capsule design or manufacturing process that may be leading to a lower reuse count.

#### SQL Functions
This query utilizes a CTE, Group By, Window and subquery function.

In [41]:
%%sql  
WITH capsule_type_counts AS (
  SELECT type, reuse_count, COUNT(*) AS capsule_count,
         ROW_NUMBER() OVER (PARTITION BY type ORDER BY reuse_count DESC) AS rn
  FROM capsules
  WHERE reuse_count IS NOT NULL
  GROUP BY type, reuse_count
)

SELECT type, AVG(reuse_count) AS avg_reuse_count,
       IF(MOD(capsule_count, 2) = 0,
          (SELECT AVG(reuse_count)
           FROM capsule_type_counts
           WHERE type = c.type AND rn IN (capsule_count / 2, capsule_count / 2 + 1)),
          (SELECT reuse_count
           FROM capsule_type_counts
           WHERE type = c.type AND rn = (capsule_count + 1) / 2)
       ) AS median_reuse_count
FROM capsule_type_counts c
WHERE rn = 1
GROUP BY type;

 * mysql://admin:***@project-data.cinjnyndseol.us-east-1.rds.amazonaws.com/API-Data
3 rows affected.


[('Dragon 1.0', Decimal('0.0000'), None),
 ('Dragon 1.1', Decimal('2.0000'), Decimal('1.0000')),
 ('Dragon 2.0', Decimal('0.0000'), None)]

### Insight and Reccomendation
This query shows that only the Dragon 1.1 capsule types are being re-used.

If the goal is to increase the sustainability of space travel by promoting the reuse of spacecraft components, the company could focus on improving the design or maintenance of the Dragon 1.1 capsule to increase its reuse count. On the other hand, they may decide to reduce the production of the Dragon 1.0 and Dragon 2.0 types if they are not being reused, to reduce waste and save resources.